Update algorithm to FSRS 4.5

SplittyDev opened the pull request:
Hey, this PR updates the algorithm to FSRS 4.5.

## Changes

- Introduced `decay` and `factor` params
- Updated forgetting curve calculation
- Simplified some of the code for clarity and deduplication
- Fixed tests to pass again

## Sources

New values for the forgetting curve etc. have been taken from:
- https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#fsrs-45
- https://github.com/open-spaced-repetition/go-fsrs/blob/main/params.go
- https://github.com/open-spaced-repetition/go-fsrs/blob/main/fsrs.go

Further, the tests have been updates to use the same values (for date and weights) as the Go tests:
- https://github.com/open-spaced-repetition/go-fsrs/blob/main/fsrs_test.go

This should make it easier to maintain both versions in the future and test them against each other.

## Misc

Resolves #10

SplittyDev commented on Sources/FSRS/FSRS.swift:
```diff
@@ -75,10 +75,8 @@ public struct Card: Equatable, Codable {
-        let elapsedDays = max(0, (now.timeIntervalSince(lastReview) / Constants.secondsInDay))
-        return exp(log(0.9) * Double(elapsedDays) / stability)
+    func forgettingCurve(params p: Params) -> Double {
+        pow(1.0 + p.factor * elapsedDays / stability, p.decay)
```
This needs to guard again `stability` being `0`, but which value should we return in that case? Since (if my understanding is correct) the `forgettingCurve` calculates the `retrievability`, I think `0` should be fine, but someone with more knowledge of the algorithm should double-check this.

ishiko732 commented on Sources/FSRS/FSRS.swift:
```diff
@@ -363,6 +368,6 @@ public struct FSRS {
 
     public func nextForgetStability(d: Double, s: Double, r: Double) -> Double {
-        p.w[11] * pow(d, -p.w[12]) * (pow(s + 1.0, p.w[13]) - 1) * exp((1 - r) * p.w[14])
+        min(p.w[11] * pow(d, -p.w[12]) * (pow(s + 1.0, p.w[13]) - 1) * exp((1.0 - r) * p.w[14]), s)
```
The min(next_s, current_s) is not necessary.
- https://open-spaced-repetition.github.io/ts-fsrs/classes/FSRSAlgorithm.html#next_forget_stability
- https://github.com/open-spaced-repetition/fsrs4anki/blob/67b09049dee058120db4ea2e021d69a58f860dff/fsrs4anki_scheduler.js#L209-L214
- https://github.com/open-spaced-repetition/go-fsrs/blob/3a13e86ea44e50ff2e33fe54b01099bcd5cddb8a/fsrs.go#L194-L199

SplittyDev commented on Sources/FSRS/FSRS.swift:
```diff
@@ -75,10 +75,8 @@ public struct Card: Equatable, Codable {
         self.lastReview = lastReview
     }
 
-    func retrievability(for now: Date) -> Double? {
```
Hmm please correct me if I'm wrong, but as far as I can tell, the `retrievability` method here was completely unused.

A quick code search confirms that:
https://github.com/search?q=repo%3Aopen-spaced-repetition%2Fswift-fsrs%20retrievability&type=code

The actual `retrievability` value was previously calculated here:
https://github.com/open-spaced-repetition/swift-fsrs/blob/0765052025215685b1fc920bb0b9bb93521ddd48/Sources/FSRS/FSRS.swift#L278

I've removed the old method and introduced a new one to properly calculate the retrievability using the same formula used in the Anki Extension and described in the docs for FSRS 4.5.

Since the elapsed days are already baked into the card and updated _before_ `forgettingCurve` is called (which only happens once anyway), I would think that this code doesn't overlook the time.

However, I implemented this stuff quite quickly without much previous knowledge of the code, so I might be overlooking something important here. I'm fairly confident that this method was completely unused though.

ishiko732 replied:
I'm not certain whether the developers using the swift-fsrs library have utilized the retrievability method, but the retrievability method has practical significance. It allows the specified card to display its retrievability at a specific time, such as the retrievability of the card tomorrow or the day after. If the retrievability method is removed, developers would have to implement the FSRS formula themselves. 
- https://github.com/open-spaced-repetition/py-fsrs/pull/9

ishiko732 replied:
I hope the `retrievability` method can be retained, and it should be able to calculate the actual `elapsed` value based on the user's input time, and then pass it to the `forgettingCurve` to calculate the R value. Of course, this `retrievability` method should only be applicable for cards in the `State.Review` state.

SplittyDev replied:
I see, thanks for the clarification! I added back the `retrievability` method.

SplittyDev commented on Sources/FSRS/FSRS.swift:
```diff
@@ -74,11 +74,16 @@ public struct Card: Equatable, Codable {
+        return forgettingCurve(elapsedDays: elapsedDays, params: params)
+    }
+
+    func forgettingCurve(elapsedDays: Double, params p: Params) -> Double {
+        guard !stability.isZero else { return 0 }
+        return pow(1.0 + p.factor * elapsedDays / stability, p.decay)
     }
```
Which value should we return in the case that `stability` is `0`?
I think `0` should be fine, but someone with more knowledge of the algorithm should double-check this.

ishiko732 replied:
Actually, s will never be 0, and this method will still be called. When the state is .new, the D and S (difficulty and stability) will be initialized.
- https://github.com/open-spaced-repetition/swift-fsrs/blob/0765052025215685b1fc920bb0b9bb93521ddd48/Sources/FSRS/FSRS.swift#L255-L256
- https://github.com/open-spaced-repetition/swift-fsrs/blob/0765052025215685b1fc920bb0b9bb93521ddd48/Sources/FSRS/FSRS.swift#L208-L212

SplittyDev replied:
That's true, but the user can call `forgettingCurve` before the card is ever properly initialized. I don't think it hurts to leave this in as a safeguard (FYI I had my app crash specifically because of this).

SplittyDev commented:
@ishiko732 I think I've addressed all of your concerns now.
Please let me know if the `0` handling for `forgettingCurve(...)` where `stability == 0` is fine.

ishiko732 reviewed (approved):
LGTM

ishiko732 commented:
@L-M-Sherlock  @bensmiley  cc

L-M-Sherlock commented on Tests/FSRSTests/FSRSTests.swift:
```diff
@@ -48,9 +51,11 @@ final class FSRSTests: XCTestCase {
+        XCTAssertEqual(statusHistory, [
+            .new, .learning, .review, .review, .review, .review, .review,
+            .relearning, .relearning, .review, .review, .review, .review
+        ])
```
Could you add this test?

https://github.com/open-spaced-repetition/go-fsrs/blob/3a13e86ea44e50ff2e33fe54b01099bcd5cddb8a/fsrs_test.go#L58-L87

SplittyDev replied:
Sure thing! Added the test, luckily it passes ✅

L-M-Sherlock reviewed (approved):
LGTM
